### ISSUE TITLE
AssistedFactory generation fails under IC

### ISSUE THREAD

**Issue by andreaswaltenberger** [square/anvil#707]
Starting with Anvil 2.4.5 (and probably more importantly Kotlin 1.8.20) the AssistedFactory generation does not happen anymore after a change with incremental compilation.

I have a repro case available here: https://github.com/square/anvil/compare/main...andreaswaltenberger:anvil:repro/assisted_factory

On the first build everything works fine, but when we remove the comments around `@Assisted("mom") mother: String` in the constructor and the Factory we run into issues when doing an incremental build.
It seems the Anvil compiler is not correctly pick up the change and no longer generates a new/adapted class - it is just missing which then fails kapt/Dagger with:
```
import com.squareup.anvil.sample.child.Child_Factory_Impl;
                                      ^
symbol:   class Child_Factory_Impl
location: package com.squareup.anvil.sample.child
```

When rerunning the build with `./gradlew :sample:app:assembleDebug --rerun-tasks` everything is compiled correctly again.

I'm not 100% sure this is an Anvil issue or coming from the change in the Kotlin compiler with the new optimisations turned on by default (`kapt.use.jvm.ir=true` or `kotlin.incremental.useClasspathSnapshot=true`).

Any ideas what could be going on? Happy to help with additional information if needed, thanks 🙏

**Comment by JoelWilcox** [comment]
Looks like this is the same issue as https://github.com/square/anvil/issues/693, which is currently caused by annotation changes not being tracked/factored in for incremental compilation. ATM it's looking like it won't be fixed until 1.9.0-Beta. Related Kotlin tickets: 

https://youtrack.jetbrains.com/issue/KT-58289/IC-fails-to-detect-a-change-to-class-annotations
https://youtrack.jetbrains.com/issue/KT-57919

**Comment by andreaswaltenberger** [comment]
Thanks for looking into this but unfortunately I think we are running into a different issue here 😕 (It might be even related to #708?)

Since 1.9.0-Beta is now "officially" out I bumped the version in my repo linked above but the outcome is unchanged.

**Comment by ZacSweers** [comment]
This is a dupe of #693

**Comment by andreaswaltenberger** [comment]
Makes sense - thanks a lot for following up guys, appreciate all the work you're putting into this. ❤️ 
Will keep an eye on the other issue 👍
